### PR TITLE
Improve session polling to prevent accidental redirects

### LINKED ISSUE #33071: RESTART_AUTHENTICATION_ERROR in Iphone devices (using safari and chrome browser)
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

oidc

### Describe the bug

Hello everyone,

I'm experiencing an issue with Keycloak when authenticating on iPhone devices (Safari and Chrome). The authentication flow works correctly on desktop browsers (Chrome on Windows), but on iPhone, the process fails, and instead of receiving the expected authorization code parameter, the authentication is restarted, or I receive a temporarily_unavailable error.

Working Flow (Desktop Browser - Chrome)

The user is redirected to Keycloak and logs in successfully.
After login, the browser receives a URL with the code= parameter, which is then used to complete the OpenID Connect flow and obtain tokens.
Everything works as expected.

Failing Flow (iPhone - Safari/Chrome)
The user is redirected to Keycloak and logs in successfully.
Instead of receiving the code= parameter, the authentication flow redirects to /login-actions/restart with a 302 status, indicating that the authentication session is being restarted.
This flow never completes, and an error temporarily_unavailable or authentication_expired appears in the logs.

### Version

25.0.1

### Regression

- [ ] The issue is a regression

### Expected behavior

After login, the browser receives a URL with the code= parameter, which is then used to complete the OpenID Connect flow and obtain tokens.

### Actual behavior

The user is redirected to Keycloak and logs in successfully.
Instead of receiving the code= parameter, the authentication flow redirects to /login-actions/restart with a 302 status, indicating that the authentication session is being restarted.

### How to Reproduce?

have an isntance of Keycloak 25.0.1 running
create a confidential client in master realm
Enable Authentication Flow
and set a redirect Uri
Test the Flow on Desktop Browser
Test the Flow on iPhone (Safari or Chrome)

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by jonkoops**
Improves the session polling in the login flow used to detect user sign-ins from other contexts (e.g. another tab or window). This fixes an issue where a user experiences a redirect to another page when submitting the login form, causing an unexpected re-start  authentication error.

This change also cleans up the code a little by renaming some functions to be more appropriate with their purpose, as well as adding additional comments to explain various parts of the code.

Closes #33071

### CODE REVIEW

**Review by edewit — DISMISSED** [review]

**Comment by jonkoops** [comment]
Putting this on hold until we're able to test that this indeed fixes the issue, please do not merge until this label is removed.

File: themes/src/main/resources/theme/base/login/resources/js/authChecker.js
Comment (graziang) [inline themes/src/main/resources/theme/base/login/resources/js/authChecker.js:38]: Is this line executed?

Comment (jonkoops) [reply themes/src/main/resources/theme/base/login/resources/js/authChecker.js:38]: Yes, starting a navigation does not prevent additional JavaScript from executing whilst the current page is being unloaded.

**Review by graziang — APPROVED** [review]

**Review by rmartinc — APPROVED** [review]
Thanks @jonkoops! Works OK to me! Let's see what @t-tani says when testing in his env...

**Review by mposolda — DISMISSED** [review]

**Comment by jonkoops** [comment]
As far as the team can tell this fix should work for the reported issue, we intend to merge this in next week Monday if there are no objections.

**Comment by t-tani** [comment]
@jonkoops I'm sorry for my late response.
I tested you PR with my environment with the following setup:

- **Server**: Keycloak 26.0.5  
- **Proxy**: BurpProxyPro (intercepting `authChecker.js` and rewriting to the PR authCheck.js)  
- **Client**: iPhone 14 Pro (iOS 18.0.1 / Safari)  
- **OpenIDC Client**: Cloudflare  

### Tested Authentication Methods and Results  
- **Password**:  OK  
- **OTP**:  OK  
- **Passkey**:  NG  

The issue persists only in the case of Passkey authentication. Upon inspecting the HTML, I found that the "Sign in with Passkey" button does not use a form submission during login. As a result, the session polling event is not properly stopped.

Here is the relevant HTML structure for Passkey authentication:

```html
<div class="pf-v5-c-form__group no-bottom-margin">
    <form id="authn_select" class="pf-v5-c-form">
        <input type="hidden" name="authn_use_chk" value="<redacted>"/>
        <input type="hidden" name="authn_use_chk" value="<redacted>"/>
        <input type="hidden" name="authn_use_chk" value="<redacted>"/>
    </form>

    <div id="kc-form-buttons" class="col-xs-12 col-sm-12 col-md-12 col-lg-12">
        <input id="authenticateWebAuthnButton" type="button" autofocus="autofocus"
               value="Sign in with Passkey"
               class="pf-v5-c-button pf-v5-c-button pf-m-primary pf-m-block btn-lg"/>
    </div>
</div>
```

To resolve the issue in my environment, I added the following code to the proposed implementation, which attaches a `stopSessionPolling()` event to all `input[type="button"]` elements:

```javascript
const buttonInputs = document.querySelectorAll('input[type="button"]');

buttonInputs.forEach((input) =>
  input.addEventListener("click", () => stopSessionPolling())
);
```

While I have not thoroughly tested the impact of attaching `stopSessionPolling()` to all `input[type="button"]` elements on other functionalities, this change successfully resolved the Passkey issue in iOS browsers during my testing.

**Comment by jonkoops** [comment]
@t-tani thank you so much for doing this detailed testing, it really helps us out! The reason why the form submission was not working is because we're calling [`.submit()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/submit) from JavaScript in various places, which does not send out a `submit` event:

> This method is similar, but not identical to, activating a form's submit [`<button>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button). When invoking this method directly, however:
> - No [`submit`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/submit_event) event is raised. In particular, the form's `onsubmit` event handler is not run.
> - [Constraint validation](https://developer.mozilla.org/en-US/docs/Web/HTML/Constraint_validation) is not triggered.
>
> The [`HTMLFormElement.requestSubmit()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/requestSubmit) method is identical to activating a form's submit [`<button>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button) and does not have these differences.

I went ahead an replaced all instances where `.submit()` was called with `.requestSubmit()`, and this seems to resolve the issue you encountered as the `submit` event is now triggered. Could you take another look to validate this?

**Comment by jonkoops** [comment]
Great to hear! I'll make sure to get this merged and backported ASAP. Thanks again for all your assistance!

**Review by rmartinc — APPROVED** [review]

**Review by graziang — APPROVED** [review]

**Review by mposolda — APPROVED** [review]

**Comment by jonkoops** [comment]
Backport for v26 is up under #35408

**Comment by FlxPeters** [comment]
@mposolda The change from `submit()` to `requestSubmit()` unfortunately introduces a breaking change for some clients.
The Microsoft Office 365 desktop clients have an embedded Edge 18 Browser, [which does not support the method](https://caniuse.com/?search=requestSubmit). 
Due to this, the login it works as expected, but breaks the MFA selector form which prevents a successful login on these clients. This was partially fixed in #35659, but only for the SAML post form. 

I would like to suggest rolling this back to the `submit()` method, where it's not a hard dependency.
It was pretty hard to find the root cause of this issue since it's not a specific change and more a side effect. If you're fine with this suggestion, I can provide an MR.

Side note: We also try to open a case at Microsoft to evaluate why they ship such an old embedded browser.

**Comment by rmartinc** [comment]
@FlxPeters You can always create a custom theme for those old clients changin the method.

**Comment by FlxPeters** [comment]
@rmartinc Sure, and we have to for now. My point is that this should be a dedicated change and not a side effect of another one. No bad feelings at all, such things happen. I opened an MR for the rollback: https://github.com/keycloak/keycloak/pull/38558

**Comment by rmartinc** [comment]
@FlxPeters If you rollback the change then the issue #33071 is again in place. Which is for iphone users. So or we have issues in one place or in another, the only real fix would be separating by browsers that we are don't want to do. Somebody should have to change themes. IMHO it should be the one not supporting the new standards. But let's see what people think.

**Pull request merged**
